core/libp2p/protocol_muxer/multiselect/stream_state.hpp

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_STREAM_STATE_HPP
+#define KAGOME_STREAM_STATE_HPP
+
+#include <functional>
+#include <optional>
+
+#include "libp2p/peer/peer_id.hpp"
+#include "libp2p/protocol_muxer/protocol_muxer.hpp"
+#include "libp2p/stream/stream.hpp"
+
+namespace libp2p::protocol_muxer {
+  /**
+   * Stores current state of protocol negotiation over the stream
+   */
+  struct StreamState {
+    enum class NegotiationStatus {
+      OPENING_SENT,
+      PROTOCOL_SENT,
+      PROTOCOLS_SENT,
+      LS_SENT,
+      NA_SENT
```

Comment:
it looks like a state machine. use sml

Reply:
Well, it's really represented by an FSA, but I don't think that using SML library for it is better, than simple enum-switch statements:
1. Most of the states are any-to-any, and transitions are relatively simple or even trivial - so no complex logic it to be modelled
2. It will increase a complexity of the code without a useful gain back - again, if the logic is complex, library variuant can help to understand the FSA better, but here enum-switches are more readable IMHO
3. Compilation time will increase
So, I don't think that SML is a good idea in this paricular case